// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py
// RUN: %clang_cc1 -triple riscv32 -target-feature +experimental-zimop -emit-llvm -verify %s -o - \
// RUN:     -disable-O0-optnone
// RUN: %clang_cc1 -triple riscv64 -target-feature +experimental-zimop -emit-llvm -verify %s -o - \
// RUN:     -disable-O0-optnone

#include <stdint.h>

#if __riscv_xlen == 64

uint64_t mopr_41_64(uint64_t a) {
  return __builtin_riscv_mopr_64(a, 41); // expected-error {{argument value 41 is outside the valid range [0, 31]}}
}

uint64_t moprr_8_64(uint64_t a, uint64_t b) {
  return __builtin_riscv_moprr_64(a, b, 8); // expected-error {{argument value 8 is outside the valid range [0, 7]}}
}

#endif

uint32_t mopr_88_32(uint32_t a) {
  return __builtin_riscv_mopr_32(a, 88); // expected-error {{argument value 88 is outside the valid range [0, 31]}}
}

uint32_t moprr_16_32(uint32_t a, uint32_t b) {
  return __builtin_riscv_moprr_32(a, b, 16); // expected-error {{argument value 16 is outside the valid range [0, 7]}}
}

